lib/phoenix_live_view/async.ex

```diff
@@ -3,11 +3,71 @@ defmodule Phoenix.LiveView.Async do
 
   alias Phoenix.LiveView.{AsyncResult, Socket, Channel}
 
-  def start_async(%Socket{} = socket, key, func, opts \\ []) when is_function(func, 0) do
-    run_async_task(socket, key, func, :start, opts)
+  defp validate_function_env(func, op, env) do
+    # TODO: is there a better way to check if socket is used in func?
+    Macro.prewalk(func, fn
+      {:socket, _, nil} ->
+        IO.warn(
+          """
+          you are accessing the LiveView Socket inside a function given to #{op}.
+
+          This is an expensive operation because the whole socket is copied to the new process.
+
+          Instead of:
+
+              #{op}(socket, :key, fn ->
+                do_something(socket.assigns.my_assign)
+              end)
+
+          You should do:
+
+              my_assign = socket.assigns.my_assign
+
+              #{op}(socket, :key, fn ->
+                do_something(my_assign)
+              end)
+
+          For more information, see https://hexdocs.pm/elixir/1.16.1/process-anti-patterns.html#sending-unnecessary-data.
+          """,
+          file: env.file,
+          line: env.line
+        )
+
+      other ->
+        other
+    end)
+
+    :ok
   end
 
-  def assign_async(%Socket{} = socket, key_or_keys, func, opts \\ [])
+  def start_async(socket, key, func, opts, env) do
+    validate_function_env(func, :start_async, env)
+
+    quote do
+      Phoenix.LiveView.Async.run_async_task(
+        unquote(socket),
+        unquote(key),
+        unquote(func),
+        :start,
+        unquote(opts)
+      )
+    end
+  end
+
+  def assign_async(socket, key_or_keys, func, opts, env) do
+    validate_function_env(func, :assign_async, env)
+
+    quote do
+      Phoenix.LiveView.Async.do_assign_async(
+        unquote(socket),
+        unquote(key_or_keys),
+        unquote(func),
+        unquote(opts)
+      )
+    end
+  end
+
+  def do_assign_async(%Socket{} = socket, key_or_keys, func, opts)
```

Comment:
This would be fine btw. Your call.

```suggestion
      Phoenix.LiveView.Async.assign_async(
        unquote(socket),
        unquote(key_or_keys),
        unquote(func),
        unquote(opts)
      )
    end
  end

  def assign_async(%Socket{} = socket, key_or_keys, func, opts)
```